File: 2023/02/solution.ml
let lines =
  Seq.of_dispenser (fun _ ->
    match read_line () with
    | x -> Some x
    | exception End_of_file -> None)
  |> Array.of_seq

type ball =
  | Red of int
  | Green of int
  | Blue of int

type shown_balls =
  { red : int
  ; green : int
  ; blue : int
  }

let to_total balls_shown =
  balls_shown
  |> List.fold_left
       (fun totals ball ->
         match ball with
         | Red count -> { totals with red = totals.red + count }
         | Green count -> { totals with green = totals.green + count }
         | Blue count -> { totals with blue = totals.blue + count })
       { red = 0; green = 0; blue = 0 }

let parse_line str =
  (* Example: "Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green" *)
  match String.split_on_char ':' str with
  | [ game; rest ] ->
    ( Scanf.sscanf game "Game %d" (fun g -> g)
    , rest
      |> String.split_on_char ';'
      |> List.map (fun turn ->
           turn
           |> String.split_on_char ','
           |> List.map (fun ball ->
                Scanf.sscanf (String.trim ball) "%d %s" (fun count colour ->
                  match colour with
                  | "red" -> Red count
                  | "green" -> Green count
                  | "blue" -> Blue count
                  | _ -> failwith "invalid colour")))
      |> List.map to_total )
  | _ -> failwith "invalid line"

let part1 () =
  let maxTotals = { red = 12; green = 13; blue = 14 } in
  lines
  |> Array.to_seq
  |> Seq.map parse_line
  |> Seq.filter_map (fun (game, totals) ->
       if totals
          |> List.for_all (fun total ->
               total.red <= maxTotals.red
               && total.green <= maxTotals.green
               && total.blue <= maxTotals.blue)
       then Some game
       else None)
  |> Seq.fold_left ( + ) 0
  |> Printf.printf "Part1: %d\n"

let part2 () =
  lines
  |> Array.to_seq
  |> Seq.map parse_line
  |> Seq.map snd
  |> Seq.map (fun shown ->
       shown
       |> List.fold_left
            (fun total shown ->
              { red = max total.red shown.red
              ; green = max total.green shown.green
              ; blue = max total.blue shown.blue
              })
            { red = 0; green = 0; blue = 0 })
  |> Seq.map (fun total -> total.red * total.green * total.blue)
  |> Seq.fold_left ( + ) 0
  |> Printf.printf "Part2: %d\n"

let () =
  part1 ();
  part2 ()

File: 2023/02/solution.exs
defmodule Part1 do
  def parse_line(line) do
    # Example: "Game 11: 3 blue, 4 red; 5 green, 6 red; 2 green"
    # Split on ":"
    [game, rest] = String.split(line, ": ")
    game = String.replace(game, "Game ", "") |> String.to_integer()

    nums =
      rest
      |> String.split("; ")
      |> Enum.map(fn x ->
        x
        |> String.split(", ")
        |> Enum.map(fn x -> x |> String.split(" ") end)
        |> Enum.reduce(%{red: 0, green: 0, blue: 0}, fn [num, color], acc ->
          case color do
            "red" -> %{acc | red: acc.red + String.to_integer(num)}
            "green" -> %{acc | green: acc.green + String.to_integer(num)}
            "blue" -> %{acc | blue: acc.blue + String.to_integer(num)}
          end
        end)
      end)

    {game, nums}
  end

  def solve(lines) do
    max = %{red: 12, green: 13, blue: 14}

    lines
    |> Enum.map(&parse_line/1)
    |> Enum.filter(fn {_game, balls} ->
      balls
      |> Enum.all?(fn nums ->
        nums.red <= max.red && nums.green <= max.green && nums.blue <= max.blue
      end)
    end)
    |> Enum.map(fn {game, _nums} ->
      game
    end)
    |> Enum.sum()
  end
end

defmodule Part2 do
  def solve(lines) do
    lines
    |> Enum.map(&Part1.parse_line/1)
    |> Enum.map(fn {_game, shown} ->
      maxes =
        shown
        |> Enum.reduce(%{red: 0, green: 0, blue: 0}, fn nums, acc ->
          %{
            red: max(acc.red, nums.red),
            green: max(acc.green, nums.green),
            blue: max(acc.blue, nums.blue)
          }
        end)

      maxes.red * maxes.green * maxes.blue
    end)
    |> Enum.sum()
  end
end

# Read all lines from stdin
lines =
  IO.read(:stdio, :eof)
  |> String.split("\n")

IO.write("Part 1: ")
Part1.solve(lines) |> IO.inspect()

IO.write("Part 2: ")
Part2.solve(lines) |> IO.inspect()
